Add player list tab in PDA

tornac1234 opened the pull request:
- [x] Add a new tab in PDA
- [x] Make player's entry
- [x] Use our own assets for the UI (need to fix the assets not fitting correctly)
- [x] Wait for #1740 to be merged before implementing more stuff
- [x] Add **documentation** especially for patches
- [x] Make the system modular
- [x] Add confirmation popups when clicking on buttons
- [x] ~~Before merging: move the translations~~ not necessary

Fixes #1475

tornac1234 commented:
Would need to know if that is what you wanted that to be made

tornac1234 commented:
Also I kept those translations but will remove them once that's ready

tornac1234 commented:
Reviews would be appreciated at this point, there's no more feature to be added

Jannify reviewed (changes requested):
Didn't review everything (most of the UI and patches code, will do that later)

Jannify commented:
Also you should upload the assets of your AssetBundle.

tornac1234 commented:
> Also you should upload the assets of your AssetBundle.

Where should I upload them ?

tornac1234 commented on NitroxClient/Unity/Helper/AssetBundleLoader.cs:
```diff
@@ -65,5 +103,30 @@ public static IEnumerator LoadUIAsset(string bundleName, bool hideUI, Action<Gam
+        return loadedBundles.Contains(bundleName);
+    }
+
+    // Equivalent to an enum with string values
```
If we use `Enum.GetName` I'd say that if we change a bundle's name, we should make the modifications everywhere while here we can simply change that property.

Jannify commented:
> > Also you should upload the assets of your AssetBundle.
> 
> Where should I upload them ?

/NitroxUnity ?
How did you create the assetbubdle in the first place?

tornac1234 commented:
> > > Also you should upload the assets of your AssetBundle.
> > 
> > 
> > Where should I upload them ?
> 
> /NitroxUnity ? How did you create the AssetBundle in the first place?

Huh, I created a unity project to be able to tests the sprites and it's from there that I built those asset bundles

tornac1234 commented on NitroxClient/Unity/Helper/AssetBundleLoader.cs:
```diff
@@ -10,40 +11,76 @@ public static class AssetBundleLoader
     private static readonly string assetRootFolder = NitroxUser.AssetsPath;
 
-    private static bool firstTime = true;
+    // Two operations should not run at the same time
```
Ok actually the problem wasn't about that but the issue is shared assets weren't loaded when the chat object spawned.

tornac1234 commented:
Just like so, the `NitroxAssetBundle` type is more practical because of the `implicit operator string` that I provided. I also added a more useful callback system for the `AssetBundleLoader`. Concerning `AssetsHelper`, I moved its functions to `uGUI_PlayerListTab`.

tornac1234 commented on NitroxClient/GameLogic/HUD/Components/uGUI_PlayerPingEntry.cs:
```diff
@@ -0,0 +1,254 @@
+    {
+        LocalPlayer localPlayer = NitroxServiceLocator.LocateService<LocalPlayer>();
+        
+        bool isNotLocalPlayer = !IsLocalPlayer || true;
```
```suggestion
        bool isNotLocalPlayer = !IsLocalPlayer;
```
Forgot to remove that testing part (useful because you can test with only one client)

tornac1234 replied:
Just letting this comment there for testing purposes (when alone) but it's no longer in the code

tornac1234 commented on NitroxClient/GameLogic/ChatUI/PlayerChatManager.cs:
```diff
@@ -96,22 +97,20 @@ public void SendMessage()
-        public void LoadChatKeyHint()
+        public IEnumerator LoadChatKeyHint()
         {
             if (!NitroxPrefs.ChatUsed.Value)
```
```suggestion
            if (!NitroxPrefs.ChatUsed.Value || true)
```
For testing purposes

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_PDA_Initialize_Patch.cs:
```diff
@@ -0,0 +1,90 @@
+        Resolve<NitroxPDATabManager>().CustomTabs.Keys.ForEach(uGUI_PDA.regularTabs.Add);
+        //
+        
+        __instance.tabs = new Dictionary<PDATab, uGUI_PDATab>(uGUI_PDA.sPDATabComparer)
+        {
+            {
+                PDATab.Intro,
+                __instance.tabIntro
+            },
+            {
+                PDATab.Inventory,
+                __instance.tabInventory
+            },
+            {
+                PDATab.Journal,
+                __instance.tabJournal
+            },
+            {
+                PDATab.Ping,
+                __instance.tabPing
+            },
+            {
+                PDATab.Gallery,
+                __instance.tabGallery
+            },
+            {
+                PDATab.Log,
+                __instance.tabLog
+            },
+            {
+                PDATab.Encyclopedia,
+                __instance.tabEncyclopedia
+            },
+            {
+                PDATab.TimeCapsule,
+                __instance.tabTimeCapsule
+            }
+        };
```
```suggestion
        __instance.tabs = new Dictionary<PDATab, uGUI_PDATab>(uGUI_PDA.sPDATabComparer)
        {
            { PDATab.Intro, __instance.tabIntro },
            { PDATab.Inventory, __instance.tabInventory },
            { PDATab.Journal, __instance.tabJournal },
            { PDATab.Ping, __instance.tabPing },
            { PDATab.Gallery, __instance.tabGallery },
            { PDATab.Log, __instance.tabLog },
            { PDATab.Encyclopedia, __instance.tabEncyclopedia },
            { PDATab.TimeCapsule, __instance.tabTimeCapsule }
        };
```
Improves readability imo.

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_PDA_Initialize_Patch.cs:
```diff
@@ -0,0 +1,90 @@
+        // Also required to add the custom tabs here
+        foreach (KeyValuePair<PDATab, NitroxPDATab> nitroxTab in Resolve<NitroxPDATabManager>().CustomTabs)
+        {
+            __instance.tabs.Add(nitroxTab.Key, nitroxTab.Value.uGUI_PDATab);
```
Can't the first foreach (L27/28) be moved here?

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_PDA_Initialize_Patch.cs:
```diff
@@ -0,0 +1,90 @@
+
+        // Initialize all the custom tabs so that they can create their required components
+        // And add their "types" to the tab list
+        Resolve<NitroxPDATabManager>().CustomTabs.Values.ForEach(tab => tab.OnInitializePDA(__instance));
+        Resolve<NitroxPDATabManager>().CustomTabs.Keys.ForEach(uGUI_PDA.regularTabs.Add);
```
```suggestion
        foreach (KeyValuePair<PDATab, NitroxPDATab> keyValuePair in Resolve<NitroxPDATabManager>().CustomTabs)
        {
            keyValuePair.Value.OnInitializePDA(__instance);
            uGUI_PDA.regularTabs.Add(keyValuePair.Key);
        }
```
Or is the double iteration needed?

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_PDA_Initialize_Patch.cs:
```diff
@@ -0,0 +1,90 @@
+    private readonly static MethodInfo TARGET_METHOD = Reflect.Method((uGUI_PDA t) => t.Initialize());
+
+    // Added stuff is surrounded by comments
+    public static bool Prefix(uGUI_PDA __instance)
```
If we are only inserting our stuff why didn't you use a transpiler?

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_PDA_SetTabs_Patch.cs:
```diff
@@ -0,0 +1,60 @@
+
+public class uGUI_PDA_SetTabs_Patch : NitroxPatch, IDynamicPatch
+{
+    private readonly static MethodInfo TARGET_METHOD = Reflect.Method((uGUI_PDA t) => t.SetTabs(default));
```
```suggestion
    private static readonly MethodInfo TARGET_METHOD = Reflect.Method((uGUI_PDA t) => t.SetTabs(default));
```

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_PDA_SetTabs_Patch.cs:
```diff
@@ -0,0 +1,60 @@
+{
+    private readonly static MethodInfo TARGET_METHOD = Reflect.Method((uGUI_PDA t) => t.SetTabs(default));
+
+    public static bool Prefix(uGUI_PDA __instance, List<PDATab> tabs)
```
If it is the same like in `uGUI_PDA_Initialize_Patch` (inserting code into base method) I would like to know why it's no transpiler and if it needs to be a Prefix I would like to have the same explaining and surrounding comments like in `uGUI_PDA_Initialize_Patch`.

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_PDA_SetTabs_Patch.cs:
```diff
@@ -0,0 +1,60 @@
+        return false;
+    }
+
+    private static void AssignSprite(uGUI_Toolbar toolbar, int index, Atlas.Sprite sprite)
```
instead of the toolbar and index you can use a `uGUI_ItemIcon` parameter and get rid of altering index problems.

tornac1234 replied:
we can't because at this point of the code, uGUI_ItemIcon aren't a thing yet

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_Pings_OnWillRenderCanvases_Patch.cs:
```diff
@@ -0,0 +1,61 @@
+/// </summary>
+public class uGUI_Pings_OnWillRenderCanvases_Patch : NitroxPatch, IDynamicPatch
+{
+    private static MethodInfo TARGET_METHOD = Reflect.Method((uGUI_Pings t) => t.OnWillRenderCanvases());
```
```suggestion
    private static readonly MethodInfo TARGET_METHOD = Reflect.Method((uGUI_Pings t) => t.OnWillRenderCanvases());
```

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_Pings_OnWillRenderCanvases_Patch.cs:
```diff
@@ -0,0 +1,61 @@
+
+        List<CodeInstruction> instructionList = instructions.ToList();
+
+        for (int i = 0; i < instructionList.Count; i++)
+        {
+            CodeInstruction instruction = instructionList[i];
```
```suggestion
        foreach (CodeInstruction instruction in instructionList)
        {
```

tornac1234 commented on NitroxClient/GameLogic/HUD/Components/uGUI_PlayerPingEntry.cs:
```diff
@@ -0,0 +1,253 @@
+    public string PlayerName => player?.PlayerName ?? string.Empty;
+    public bool IsLocalPlayer => player is LocalPlayer;
+    private bool showPing;
+    private bool _muted;
```
was just for testing purposes

tornac1234 commented on NitroxClient/GameLogic/HUD/Components/uGUI_PlayerPingEntry.cs:
```diff
@@ -0,0 +1,253 @@
+        NitroxServiceLocator.LocateService<MutePlayerProcessor>().OnPlayerMuted += (playerId, muted) =>
+        {
+            if ((player is RemotePlayer remotePlayer && remotePlayer.PlayerId == playerId) ||
+                (player is LocalPlayer localPlayer && localPlayer.PlayerId == playerId))
```
was just for testing purposes

tornac1234 commented on NitroxClient/GameLogic/HUD/PlayerListTab.cs:
```diff
@@ -0,0 +1,39 @@
+
+    public override string TabIconAssetName => "player_list_tab@3x";
+    
+    public override PDATab FallbackTabIcon => PDATab.Inventory;
```
I'm pretty sure it's better to have a way to differentiate the tabs we create so I'd let it as it is

Jannify replied:
ok

Jannify commented on Nitrox.Assets.Subnautica/LanguageFiles/fr.json:
```diff
@@ -1,47 +1,62 @@
 ﻿{
-    "Nitrox_AddServer": "Ajouter un serveur",
```
Pls revert the indent changes as weblate will probably get issues with that.

Jannify commented on NitroxClient/GameLogic/HUD/PdaTabs/uGUI_PlayerListTab.cs:
```diff
@@ -0,0 +1,218 @@
+    private readonly Dictionary<string, Sprite> assets = new();
+    public bool FinishedLoadingAssets { get; private set; }
+
+    private new Dictionary<int, uGUI_PlayerPingEntry> entries = new();
+    private new List<uGUI_PlayerPingEntry> pool = new();
+    private new Dictionary<string, uGUI_PlayerPingEntry> tempSort = new();
```
```suggestion
    private new readonly Dictionary<int, uGUI_PlayerPingEntry> entries = new();
    private new readonly List<uGUI_PlayerPingEntry> pool = new();
    private new readonly Dictionary<string, uGUI_PlayerPingEntry> tempSort = new();
```

Jannify commented on NitroxClient/GameLogic/HUD/PdaTabs/uGUI_PlayerListTab.cs:
```diff
@@ -0,0 +1,218 @@
+
+        foreach (KeyValuePair<int, INitroxPlayer> entry in players)
+        {
+            int key = entry.Key;
+            INitroxPlayer player = entry.Value;
+            if (!entries.ContainsKey(key))
+            {
+                // Sets up a new entry for the player
+                AddNewEntry(entry.Key, entry.Value);
+            }
```
```suggestion
            if (!entries.ContainsKey(entry.Key))
            {
                // Sets up a new entry for the player
                AddNewEntry(entry.Key, entry.Value);
            }
```

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_PDA_Initialize_Patch.cs:
```diff
@@ -0,0 +1,60 @@
+namespace NitroxPatcher.Patches.Dynamic;
+
+/// <summary>
+/// Add custom tabs to the 
```
I think something is missing here xD

Jannify commented on NitroxPatcher/Patches/Dynamic/uGUI_PDA_SetTabs_Patch.cs:
```diff
@@ -0,0 +1,70 @@
+        }
+    }
+
+    public static void SetupNitroxIcons(uGUI_PDA __instance, List<PDATab> tabs, Atlas.Sprite[] array)
```
`List<PDATab> tabs` isn't used in the method.
